app/server/lib/ExportTableSchema.ts

```diff
@@ -0,0 +1,151 @@
+import * as express from 'express';
+import {ApiError} from 'app/common/ApiError';
+import {WidgetOptions} from 'app/common/WidgetOptions';
+import {ActiveDoc} from 'app/server/lib/ActiveDoc';
+import {DownloadOptions, exportTable} from 'app/server/lib/Export';
+
+interface ExportColumn {
+  id: number;
+  colId: string;
+  label: string;
+  type: string;
+  widgetOptions: WidgetOptions;
+  description?: string;
+  parentPos: number;
+}
+
+interface FrictionlessFormat {
+  name: string;
+  title: string;
+  schema: {
+    fields: {
+      name: string;
+      type: string;
+      description?: string;
+      format?: string;
+      bareNumber?: boolean;
+      groupChar?: string;
+      decimalChar?: string;
+      gristFormat?: string;
+      constraint?: {};
+      trueValue?: string[];
+      falseValue?: string[];
+    }[]
+  }
+}
+
+/**
+ * Return a table schema for frictionless interoperability
+ *
+ * See https://specs.frictionlessdata.io/table-schema/#page-frontmatter-title for spec
+ * @param {Object} activeDoc - the activeDoc that the table being converted belongs to.
+ * @param {Object} options - options to get the table ID
+ * @return {Promise<FrictionlessFormat>} Promise for the resulting schema.
+ */
+export async function collectTableSchemaInFrictionlessFormat(
+  activeDoc: ActiveDoc,
+  req: express.Request,
+  options: DownloadOptions
+): Promise<FrictionlessFormat> {
+  const {tableId} = options;
+  if (!activeDoc.docData) {
+    throw new Error('No docData in active document');
+  }
+
+  // Look up the table to make a CSV from.
+  const settings = activeDoc.docData.docSettings();
+  const tables = activeDoc.docData.getMetaTable('_grist_Tables');
+  const tableRef = tables.findRow('tableId', tableId);
+
+  if (tableRef === 0) {
+    throw new ApiError(`Table ${tableId} not found.`, 404);
+  }
+
+  const data = await exportTable(activeDoc, tableRef, req);
+  const tableSchema = columnsToTableSchema(tableId, data, settings.locale);
+  return tableSchema;
+}
+
+function columnsToTableSchema(
+  tableId: string,
+  {tableName, columns}: {tableName: string, columns: ExportColumn[]},
+  locale: string,
+): FrictionlessFormat {
+  return {
+    name: tableId.toLowerCase().replace(/_/g, '-'),
```

Comment:
Is this a convention of the format? It concerns me a little that the original `tableId` identifier is not recoverable from this payload.

Reply:
For `name`, frictionless does not accept `_` no uppercase. For this, we have put the key `title`.

Reply:
the `tableName` isn't necessarily the `tableId` though. Still, I won't insist on the point, it may not matter in actual use, for example if the original url is available (which has the tableId in it).